src/planscape/forsys/generate_projects_for_a_single_scenario.R

```diff
@@ -5,84 +5,98 @@ suppressMessages({
+  priority_weights,
+  stand_id_field,
+  proj_id_field,
+  stand_area_field,
+  stand_cost_field,
+  geo_wkt_field,
+  output_scenario_name,
+  output_scenario_tag,
+  PreForsysClusterType = 'KMEANS') {
+
+  # Enables debug mode if providing output_scenario_name and output_scenario_tag
   # If enabled, data and graphs are output to directory,
   # output/<output_scenario_name>/<output_scenario_tag>/
-  enable_debug <- nchar(output_scenario_name) > 0 && 
-    nchar(output_scenario_tag) > 0
-
-  priority_impact_fields <- paste(priorities, "_PCP", sep="")
+  enable_debug <- (
+    nchar(output_scenario_name) > 0 &
+    nchar(output_scenario_tag) > 0)
 
   # Appends spm and pcp data columns for each priority.
   forsys_input_data <- forsys_input_data %>% 
-                forsys::calculate_spm(fields = priorities) %>%
-                forsys::calculate_pcp(fields = priorities)
+    forsys::calculate_spm(fields = priorities) %>%
+    forsys::calculate_pcp(fields = priorities)
+
   # Appends a preset_priority column, which contains total impact score values.
+  priority_impact_fields <- paste0(priorities, "_PCP")
+  wp_colname <- "weighted_priorities"
   suppressMessages(
     forsys_input_data <- forsys_input_data %>%
-                  forsys::combine_priorities(
-                    fields = priority_impact_fields, 
-                    weights = priority_weights, 
-                    new_field = wp_str)
-  )
+      forsys::combine_priorities(
+        fields = priority_impact_fields,
+        weights = priority_weights,
+        new_field = wp_colname))
+
   # Parses wkt in the geo_wkt column and adds it to a "geometry" column.
-  # Patchmax expects column name to be "geometry" - do not change the variable
-  # name.
-  # Internally, Patchmax calls st_sf on geometry column values. This results in
-  # an error unless st_as_sfc is called on the wkt first.
-  geometry <- lapply(forsys_input_data[geo_wkt_field], st_as_sfc)
-  forsys_input_data <- cbind(forsys_input_data, geometry)
+  forsys_input_data_formatted <- forsys_input_data %>%
+    rename("geometry" = geo_wkt_field) %>%
+    st_as_sf(wkt = "geometry")
 
-  shp <- st_as_sf(forsys_input_data)
+  # k-means clustering
+  if (PreForsysClusterType == 'KMEANS') {
+    # k-means clustering execution
+    source("forsys/spatial_clustering.R")
+    forsys_input_data_formatted <- cluster_cells_to_stands(
+      stand_data = forsys_input_data_formatted,
+      scenario_priorities = c(wp_colname, priorities))
+  }
 
   # TODO: optimize project area generation parameters, SDW, EPW, sample_frac.
-  suppressMessages (
+  suppressMessages(
     run_outputs <- forsys::run(
       return_outputs = TRUE,
       write_outputs = enable_debug,
-      scenario_name = output_scenario_name,
-      scenario_write_tags = output_scenario_tag,
-      stand_data = shp,
+      stand_data = forsys_input_data_formatted,
+      stand_area_field = stand_area_field,
+      # TODO: do we want ownership or operability here?
```

Comment:
We don't want to do ownership and operability here. Please remove.

It'll be easier from python since thresholding syntax in patchmax is more limited.

Reply:
Makes sense.
Done.